src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.SignalR.Internal;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.AspNetCore.SignalR.Tests.Internal
+{
+    public class DefaultHubDispatcherTests
+    {
+        private class MockHubConnectionContext<TValue> : HubConnectionContext
+        {
+            public TaskCompletionSource<object> ReceivedCompleted = new TaskCompletionSource<object>();
+            public List<TValue> Values = new List<TValue>();
+
+            public MockHubConnectionContext(ConnectionContext connectionContext, HubConnectionContextOptions contextOptions, ILoggerFactory loggerFactory)
+                : base(connectionContext, contextOptions, loggerFactory) { }
+
+            public override ValueTask WriteAsync(HubMessage message, CancellationToken cancellationToken)
+            {
+                if (message is StreamItemMessage streamItemMessage)
+                    Values.Add((TValue)streamItemMessage.Item);
+
+                else if (message is CompletionMessage completionMessage)
+                {
+                    ReceivedCompleted.TrySetResult(null);
+
+                    if (!string.IsNullOrEmpty(completionMessage.Error))
+                    {
+                        throw new Exception("Error invoking hub method: " + completionMessage.Error);
+                    }
+                }
+
+                else throw new NotImplementedException();
+
+                return default;
+            }
+        }
+
+        private static DefaultHubDispatcher<THub> CreateDispatcher<THub>() where THub : Hub
```

Comment:
I based the instantiation of our target (the dispatcher) on the way DefaultHubDispatcherBenchmark does it. I believe this is the minimum we require to put it under test.

Reply:
We have tests that do method dispatching in [HubConnectionHandlerTests.cs](https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs) using a fake connection. I'd think these kinds of tests can go in that.

Reply:
I've moved this test into HubConnectionHandlerTests and copied the patterns used in the other tests of dispatching for streaming hubs.

I'm referring to this test case as supporting 'derived parameters' in hub methods, so I've prefixed the related artefacts (classes, interfaces, methods) with 'DerivedParameter'.
If there's another way you'd like to refer to this, let me know and I'll update the names.

Reply:
Very nice! Thanks for doing that!

I think 'DerivedParameter' is a fine name. I'll let the team give a review and expect we'll be able to get this merged in.